controller/tests/array_action/ds8k/test_array_mediator_ds8k.py

```diff
@@ -476,90 +467,118 @@ def test_get_snapshot_no_pool_success(self):
+                                                               pool_id='different_pool', tp='none')
+
+    def _test_create_snapshot_with_space_efficiency_success(self, source_volume_space_efficiency,
+                                                            space_efficiency_called, space_efficiency_parameter=None):
+        self._prepare_mocks_for_create_snapshot(tp=source_volume_space_efficiency)
+
+        if space_efficiency_parameter is not None:
+            self.array.create_snapshot("volume_id", "target_volume", space_efficiency=space_efficiency_parameter)
+        else:
+            self.array.create_snapshot("volume_id", "target_volume")
+
+        self.client_mock.create_volume.assert_called_with(name='target_volume', capacity_in_bytes=1073741824,
+                                                          pool_id='fake_pool', tp=space_efficiency_called)
+
+    def test_create_snapshot_with_specified_source_volume_space_efficiency_success(self):
+        self._test_create_snapshot_with_space_efficiency_success(source_volume_space_efficiency="none",
+                                                                 space_efficiency_called="none")
+
+    def test_create_snapshot_with_different_request_parameter_space_efficiency_success(self):
+        self._test_create_snapshot_with_space_efficiency_success(source_volume_space_efficiency="none",
+                                                                 space_efficiency_called="ese",
+                                                                 space_efficiency_parameter="thin")
+
+    def test_create_snapshot_with_different_request_parameter_empty_space_efficiency_success(self):
+        self._test_create_snapshot_with_space_efficiency_success(source_volume_space_efficiency="ese",
+                                                                 space_efficiency_called="none",
+                                                                 space_efficiency_parameter="")
 
     def test_create_snapshot_not_valid(self):
-        volume = self._prepare_mocks_for_create_snapshot()
-        self.client_mock.create_volume.return_value = volume
-        self.client_mock.get_volume.return_value = volume
+        self._prepare_mocks_for_create_snapshot()
         flashcopy_response = Munch(
             {"sourcevolume": {"id": "0001"},
              "targetvolume": {"id": "0002"},
              "id": "0001:0002",
              "state": "invalid"
              })
-        self.client_mock.create_flashcopy.return_value = flashcopy_response
-        with self.assertRaises(ValueError):
+        self.client_mock.get_flashcopies.return_value = flashcopy_response
+        with self.assertRaises(ValueError) as ar_context:
             self.array.create_snapshot("volume_id", "target_volume", pool=self.volume_response.pool)
+        self.assertIn("invalid", str(ar_context.exception))
```

Comment:
are we missing a custom exception?

Reply:
Im not sure why its `ValueError` but there is a custom msg there.

Reply:
I guess it's not for this ticket